pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+pub type UserName = Vec<u8>;
+
+pub const NAME_MAX_LEN: usize = 63;
+pub const NAME_MIN_LEN: usize = 3;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::{traits::StorageVersion, PalletId};
+    // use manta_primitives::assets::AssetConfig;
+
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Pallet ID
+        type PalletId: Get<PalletId>;
+
+        type RegisterWaitingPeriod: Get<Self::BlockNumber>;
+
+        //type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Username exists
+        NameExists,
+        /// Username not registered
+        NotRegistered,
+        /// The Registration time not reached
+        RegisterTimeNotReach,
+        /// Username invalid
+        ValidationFailed,
+        /// Already Pending Register
+        AlreadyPendingRegister,
+
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NameRegistered,
+    }
+
+    ///
+    #[pallet::storage]
+    #[pallet::getter(fn username_records)]
+    pub type UsernameRecords<T: Config> =
+        StorageMap<_, Twox64Concat, UserName, T::AccountId, OptionQuery>;
+
+    ///
+    #[pallet::storage]
+    #[pallet::getter(fn pending_register)]
+    pub type PendingRegister<T: Config> =
+        StorageMap<_, Twox64Concat, (T::Hash, T::Hash), T::BlockNumber, OptionQuery>;
+
+    ///
+    #[pallet::storage]
+    #[pallet::getter(fn primary_records)]
+    pub type PrimaryRecords<T: Config> =
+        StorageMap<_, Twox64Concat, T::AccountId, UserName, OptionQuery>;
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        ///
+        #[pallet::call_index(0)]
+        #[pallet::weight(1000)]
+        #[transactional]
+        pub fn register(
+            origin: OriginFor<T>,
+            username: Vec<u8>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            Self::do_register(&username, who)
+        }
+
+        ///
+        #[pallet::call_index(1)]
+        #[pallet::weight(1000)]
+        #[transactional]
+        pub fn accept_register(
+            origin: OriginFor<T>,
+            username: Vec<u8>,
+            price: Balance,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            Self::do_accept_register(&username, who, price)?;
+
+            // TODO: usernames as NFTs making them tradeble
+            // Self::mint_username_as_nft(&username);
```

Comment:
plan to use third party nft or use your own scheme?

Reply:
This is for now on hold, as victor wants only the username functionality